Mathlib/Analysis/Calculus/Deriv/Shift.lean

```diff
@@ -1,48 +1,63 @@
 
-We show that if a function `h` has derivative `h'` at a point `a + x`, then `h (a + ·)`
-has derivative `h'` at `x`. Similarly for `x + a`.
+We show that if a function `f` has derivative `f'` at a point `a + x`, then `f (a + ·)`
+has derivative `f'` at `x`. Similarly for `x + a`.
 -/
 
+variable {𝕜 F : Type*} [NontriviallyNormedField 𝕜] [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  {f : 𝕜 → F} {f' : F}
+
+/-- Translation in the domain does not change the derivative. -/
+lemma HasDerivAt.comp_const_add (a x : 𝕜) (hf : HasDerivAt f f' (a + x)) :
+    HasDerivAt (fun x ↦ f (a + x)) f' x := by
+  simpa [Function.comp_def] using HasDerivAt.scomp (𝕜 := 𝕜) x hf <| hasDerivAt_id' x |>.const_add a
+
+/-- Translation in the domain does not change the derivative. -/
+lemma HasDerivAt.comp_add_const (x a : 𝕜) (hf : HasDerivAt f f' (x + a)) :
+    HasDerivAt (fun x ↦ f (x + a)) f' x := by
+  simpa [Function.comp_def] using HasDerivAt.scomp (𝕜 := 𝕜) x hf <| hasDerivAt_id' x |>.add_const a
+
 /-- Translation in the domain does not change the derivative. -/
-lemma HasDerivAt.comp_const_add {𝕜 : Type*} [NontriviallyNormedField 𝕜] (a x : 𝕜) {𝕜' : Type*}
-    [NormedAddCommGroup 𝕜'] [NormedSpace 𝕜 𝕜'] {h : 𝕜 → 𝕜'} {h' : 𝕜'}
-    (hh : HasDerivAt h h' (a + x)) :
-    HasDerivAt (fun x ↦ h (a + x)) h' x := by
-  simpa [Function.comp_def] using HasDerivAt.scomp (𝕜 := 𝕜) x hh <| hasDerivAt_id' x |>.const_add a
+lemma HasDerivAt.comp_const_sub (a x : 𝕜) (hf : HasDerivAt f f' (a - x)) :
+    HasDerivAt (fun x ↦ f (a - x)) (-f') x := by
+  simpa [Function.comp_def] using HasDerivAt.scomp (𝕜 := 𝕜) x hf <| hasDerivAt_id' x |>.const_sub a
 
 /-- Translation in the domain does not change the derivative. -/
-lemma HasDerivAt.comp_add_const {𝕜 : Type*} [NontriviallyNormedField 𝕜] (x a : 𝕜) {𝕜' : Type*}
-    [NormedAddCommGroup 𝕜'] [NormedSpace 𝕜 𝕜'] {h : 𝕜 → 𝕜'} {h' : 𝕜'}
-    (hh : HasDerivAt h h' (x + a)) :
-    HasDerivAt (fun x ↦ h (x + a)) h' x := by
-  simpa [Function.comp_def] using HasDerivAt.scomp (𝕜 := 𝕜) x hh <| hasDerivAt_id' x |>.add_const a
+lemma HasDerivAt.comp_sub_const (x a : 𝕜) (hf : HasDerivAt f f' (x - a)) :
+    HasDerivAt (fun x ↦ f (x - a)) f' x := by
+  simpa [Function.comp_def] using HasDerivAt.scomp (𝕜 := 𝕜) x hf <| hasDerivAt_id' x |>.sub_const a
+
+variable (f) (a x : 𝕜)
 
 /-- The derivative of `x ↦ f (-x)` at `a` is the negative of the derivative of `f` at `-a`. -/
-lemma deriv_comp_neg {𝕜 : Type*} [NontriviallyNormedField 𝕜] {F : Type*} [NormedAddCommGroup F]
-    [NormedSpace 𝕜 F] (f : 𝕜 → F) (a : 𝕜) : deriv (fun x ↦ f (-x)) a = -deriv f (-a) := by
-  by_cases h : DifferentiableAt 𝕜 f (-a)
-  · simpa only [deriv_neg, neg_one_smul] using deriv.scomp a h (differentiable_neg _)
-  · rw [deriv_zero_of_not_differentiableAt (mt differentiableAt_comp_neg_iff.mpr h),
-      deriv_zero_of_not_differentiableAt h, neg_zero]
+lemma deriv_comp_neg : deriv (fun x ↦ f (-x)) x = -deriv f (-x) := by
+  by_cases f : DifferentiableAt 𝕜 f (-x)
```

Comment:
Shouldn't all these `deriv` lemmas be simp lemmas?

Reply:
I doubt so. Since lambdas aren't (yet) indexed in the discrimination tree, their key would all be of the form `deriv ?_ ?_` which is rather generic. Maybe not a performance issue since there aren't so many uses of `deriv`, but for now I would rather not.

Reply:
Fair enough.